rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,429 @@
+   - This could occur in the middle of a `tsh` command since we don't have control over when the touch is cached/expires.
+
+The user can provide `--piv-pin-policy=<never|always|once>` and `--piv-touch-policy=<never|always|cached>` to specify their desired policies.
+
+#### PIV slots
+
+PIV cards are required to have 4 primary slots, as well as the 20 retired key management slots. 
+
+Each slot has a different intended function and different default PIN and touch policies. Since we are manually setting the PIN and touch policies ourselves, we can technically use any of the 24 slots available in the same way.
+
+We will use slot `9a: PIV Authentication` by default, as it fits the teleport use case most closely. However, users will be able to provide their own desired slot with the flag `--piv-slot`, in case they are already using slot `9a` for a different login session or for a different PIV application, such as `yubikey-agent`.
+
+Note: Yubikeys also have an attestation slot, `f9`, which is not standardized by the PIV specifications.
+
+#### tsh PIV login session
+
+Usually, `tsh` checks the user's `~/.tsh/keys/proxy.example.com` for an existing public-private key pair. If found, that key pair will be used alongside any stored certificates to handle `tsh` calls.
+
+With PIV login session's we need a similar way to detect whether the user has an existing login session. In order to make PIV login a purely opt in feature, we will depend on information written to the user's tsh profile to determine whether the login session is PIV or non-PIV. Specifically, we will add a new `piv` field to the user's `~/.tsh/proxy.yaml` file which can also uniquely identify the PIV card and Slot being used for the login session.
+
+```yaml
+piv:
+  name: YubiKey 5 Nano (5.2.7) [OTP+FIDO+CCID] # PIV card name
+  id: xxxxxxxx                                 # unique ID
+  slot: 9a                                     # PIV slot
+```
+
+Note: for yubikey, the unique ID will be the card's serial number. Different implementations of PIV may have alternative methods for identifying unique cards
+
+##### Check the public key
+
+It's possible that after login, the user overwrites the PIV slot used to login. To catch this, future calls to `tsh` will retrieve the public key from the PIV slot and check if it matches the public key on the user's TLS certificate (`~/.tsh/keys/proxy/user-x509.pem`). If it doesn't match, then the user will see an error and get prompted to relogin.
+
+```
+> tsh ls
+ERROR: PIV private key does not match login certificates.
+// Trigger relogin
+```
+
+##### Concurrent login sessions
+
+`tsh` should support multiple concurrent login sessions for different proxies. This can be done by specifying a different slot for each concurrent login session, allowing up to 24 concurrent PIV login sessions. However, this requires that each user maintain a mental map of which slots are in use to avoid overwriting an existing login session.
```

Comment:
I originally wanted to do this, but after drafting it out it, it added unnecessary complexity:
 - If a user wants to reuse a private key slot, we need to make sure that private key passes assertion enforcement for all proxies. This becomes much more confusing for the user, and we'd need to have more granular error messages to guide the user when this issue occurs.
 - In order to track a slot's usage by `tsh`, we add a self signed certificate with login session data. If we want to track multiple login sessions with a single slot, this strategy doesn't work as well. It introduces potential race conditions between login sessions, logouts, etc.
 - Attestations stored in the backend will be set to expire when associated certificates are going to expire. This way we can regenerate new PIV private keys on every login and not worry about attestations being stored indefinitely. If a private key is reusable instead, then managing attestation storage is not as straightforward.

These issues can of course be handled, but I'm not sure it would even result in better UX.

Reply:
That's fair. Would you mind adding those to an "Alternatives Considered" section, so we don't lose track of your research?